Ios 5588 make algorand network layer

skibinalexander opened the pull request:
Сетевой слой + взаимодействие с сетевым слоем для Algorand

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5587-make-algorand-wallet-manager' into IOS-5588-make-algorand-network-layer

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdk/Common/Factories/ExternalLinkProviderFactory.swift
#	BlockchainSdk/Common/WalletAssembly/AlgorandWalletAssembly.swift
#	BlockchainSdk/WalletManagers/Algorand/AlgorandTransactionBuilder.swift
#	BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5589-make-transaction-builder-algorand' into IOS-5588-make-algorand-network-layer

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdk/Common/WalletAssembly/AlgorandWalletAssembly.swift
#	BlockchainSdk/WalletManagers/Algorand/AlgorandTransactionBuilder.swift
#	BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5589-make-transaction-builder-algorand' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 Add new models for network

skibinalexander committed:
IOS-5588 Fix changes send transaction

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5589-make-transaction-builder-algorand' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 Refactor

skibinalexander committed:
IOS-5588 Refactor

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5589-make-transaction-builder-algorand' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer

m3g0byt3 commented on BlockchainSdk/BlockchainService/Algorand/Provider/AlgorandProviderTarget.swift:
```diff
@@ -0,0 +1,84 @@
+            headers["Content-Type"] = "application/json"
+        }
+        
+        if case .nownodes = node.type, let headerName = node.apiKeyHeaderName {
+            headers[headerName] = node.apiKeyValue
+        }
```
А это получается рекомендованный самими nownodes способ передать ключ для всех сеток, не только для алго?
Не видел такого просто в остальных сетях

Для других сеток я вижу ключ просто в самом урле (xlm, near) (что менее безопасно офк)

m3g0byt3 commented on BlockchainSdk/Common/WalletAssembly/AlgorandWalletAssembly.swift:
```diff
@@ -11,15 +11,33 @@ import TangemSdk
 struct AlgorandWalletAssembly: WalletManagerAssembly {
     func make(with input: WalletManagerAssemblyInput) throws -> WalletManager {
+        let providers = [
+//            AlgorandNetworkProvider(
```
что-то закоменченное пролезло)

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
 import Foundation
 import Combine
 
+import Foundation
+import Combine
```
тоже что-то не то

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+
+private extension AlgorandWalletManager {
+    func update(with accountModel: AlgorandResponse.Account, completion: @escaping (Result<Void, Error>) -> Void) {
+        let blanaceValues = calculateCoinValueWithReserveDeposit(from: accountModel)
```
nit: typo

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+        wallet.add(coinValue: blanaceValues.coinValue)
+        wallet.add(reserveValue: blanaceValues.reserveValue)
+        
+        guard accountModel.amount >= accountModel.minBalance else {
```
Кажется с точки зрения логики странно, что мы тут сравниваем не те значения баланса, которые мы добавили в воллет на строчках выше (т.е. значения которые будут в UI видны и учитываться логикой), а значения баланса в самой сетевой модели

правильнее сравнивать `wallet.amounts[.coin] >= accountModel.minBalance` или что-то такое, так как во всей логике в аппе будет использоваться именно то что мы положили в воллет (`wallet.amounts[.coin]`)

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+    }
+    
+    private func calculateCoinValueWithReserveDeposit(from accountModel: AlgorandResponse.Account) -> (coinValue: Decimal, reserveValue: Decimal) {
+        let changeBalanceValue = accountModel.amount > accountModel.minBalance ? accountModel.amount - accountModel.minBalance : 0
```
`max(accountModel.amount - accountModel.minBalance, 0)` достаточно

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+                let sourceFee = Decimal(response.fee) / walletManager.wallet.blockchain.decimalValue
+                let minFee = Decimal(response.minFee) / walletManager.wallet.blockchain.decimalValue
+                
+                let targetFee = sourceFee > minFee ? sourceFee : minFee
```
тоже кажется можно обойтись max

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+                    genesisId: transactionInfoParams.genesisId,
+                    genesisHash: transactionInfoParams.genesisHash,
+                    firstRound: transactionInfoParams.lastRound,
+                    lastRound: transactionInfoParams.lastRound + 1000,
```
А это значение откуда, из официального воллета?

И еще напрашивается или коммент сразу тут или вынос в константы и коммент там, сложно спустя пару месяцев искать логику почему выбран тот или иной мэджикнамбер

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+                    genesisHash: transactionInfoParams.genesisHash,
+                    firstRound: transactionInfoParams.lastRound,
+                    lastRound: transactionInfoParams.lastRound + 1000,
+                    nonce: (transaction.params as? AlgorandTransactionParams)?.nonce
```
nit: если без nonce транза не уйдет, то удобнее анврапить nonce где-то и кидать ошибку если что не так
чтобы дебажить проще было

skibinalexander replied:
> nit: если без nonce транза не уйдет, то удобнее анврапить nonce где-то и кидать ошибку если что не так чтобы дебажить проще было

Без nonce транзакции нормально уходит

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
+                return buildParams
+            }
+            .withWeakCaptureOf(self)
+            .tryMap { walletManager, buildParams -> (hash: Data, params: AlgorandBuildParams) in
```
на этой строчке явный возвращаемый тип можно грохнуть будет как только GH раннер перейдет на xcode 14.3 (ща 14.2 и на нем как раз эту строчку тайпчек не хавает)

m3g0byt3 commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -9,32 +9,172 @@
     
-    func send(_ transaction: Transaction, signer: TransactionSigner) -> AnyPublisher<TransactionSendResult, Error> {
-        return .anyFail(error: WalletError.failedToSendTx)
+    private func makeNoAccountError(using accountModel: AlgorandResponse.Account) -> WalletError {
```
Мб вынести это как factory метод для WalletError?
Раз как минимум в двух местах уже юзается, NEAR и тут

skibinalexander replied:
> Мб вынести это как factory метод для WalletError? Раз как минимум в двух местах уже юзается, NEAR и тут

Не хотел бы рефакторинг тащить, у нас много где методы переиспользуются, лучше может как-то отдельно такое смотреть

skibinalexander committed:
IOS-5588 Fix review comments

skibinalexander committed:
IOS-5588 Fix blockchain coding curve parameters

skibinalexander committed:
IOS-5588 Fix review comments

skibinalexander commented on BlockchainSdk/BlockchainService/Algorand/AlgorandNetworkService.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    // MARK: - Implementation
+    
+    func getAccount(address: String) -> AnyPublisher<AlgorandResponse.Account, Error> {
```
> Вообще с архитектурной точки зрения DTO (AlgorandResponse*) не должен возвращаться сетевым сервисом и должен быть ограничен сетевым провайдером. 
> 
> Сетевой сервис принимает и возвращает уже доменные модели
> 
> 
> 
> Потому что DTO содержит разные детали имплементации сетевой части, например:
> 
> - структуру полей как задано в апишке, а не как удобно/принято в домене
> 
> - типы как задано в апишке, а не как удобно/принято в домене
> 
> - конформанс codable
> 
> - и т.д. 
> 
> 
> 
> которые все лишние и не нужны в домене
> 
> 
> 
> ---
> 
> 
> 
> Ну и на странность в архитектурном плане указывает пустота самого сетевого сервиса, сейчас он просто по сути прокся чтобы дернуть сетевой провайдер и ничего особо не делает
> 
> 

Обсудили , сделал так , как решили

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/Blockchain.swift

m3g0byt3 commented on BlockchainSdk/BlockchainService/Algorand/AlgorandNetworkService.swift:
```diff
@@ -46,10 +71,43 @@ class AlgorandNetworkService: MultiNetworkProvider {
+                .tryMap { response in
+                    /// This paramenter mast be writen for building transaction
+                    guard let genesisHash = Data(base64Encoded: response.genesisHash) else {
+                        throw WalletError.failedToBuildTx
```
nit: по смыслу тут все-таки другая ошибка должна кидаться, failedToParse
тк тут именно парсинг и валидация ответа из сети

m3g0byt3 commented on BlockchainSdk/BlockchainService/Algorand/AlgorandNetworkService.swift:
```diff
@@ -12,29 +12,54 @@ import Combine
+                        minFee: Amount(with: service.blockchain, value: minFee),
+                        fee: Amount(with: service.blockchain, value: sourceFee)
+                    )
+                    
```
тут лишние пробелы/переносы)
можно наверное просто сделать
```
return AlgorandEstimatedFeeParams(
    minFee: Amount(with: service.blockchain, value: minFee),
    fee: Amount(with: service.blockchain, value: sourceFee)
)
```

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 Fix review comments

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 Make public structure

tureck1y commented on BlockchainSdk/WalletManagers/Algorand/AlgorandWalletManager.swift:
```diff
@@ -10,31 +10,137 @@ import Foundation
+            let error = makeNoAccountError(using: accountModel)
+            completion(.failure(error))
+            return
+        }
```
Тут еще должна быть проверка и удаление пендинг транзакций

skibinalexander replied:
Планировал в рамках другого МР , дополню в этом , нет проблем

tureck1y replied:
Ну надо было тогда Туду оставить, мне же непонятно)

tureck1y commented on BlockchainSdk/BlockchainService/Algorand/DTO/AlgorandResponse.swift:
```diff
@@ -0,0 +1,11 @@
+
+import Foundation
+
+enum AlgorandResponse {}
```
потом дополнится?

skibinalexander replied:
Тут namespace , сами модельки лежат под namespace

tureck1y commented on BlockchainSdk/BlockchainService/Algorand/Provider/AlgorandProviderUrl.swift:
```diff
@@ -0,0 +1,46 @@
+    case getblock
+}
+
+struct AlgorandProviderNode: HostProvider {
```
У нас такое есть уже у нескольких сеток есть, надо подумать можно ли как-то унифицировать подход. МОжно в ноушен вынести пока

skibinalexander replied:
Планировал в рамках другого МР , дополню в этом , нет проблем

skibinalexander replied:
Закину в виде таски , помечу что нужно обсудить

skibinalexander replied:
https://tangem.atlassian.net/browse/IOS-5779

skibinalexander committed:
IOS-5588 Make clear pending transaction

skibinalexander commented:
Все-таки еще потестирую сброс pending, есть кое-какие сомнения ...

skibinalexander commented:
> Все-таки еще потестирую сброс pending, есть кое-какие сомнения ...

Отпишу как готово будет

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer

skibinalexander committed:
IOS-5588 append testnet url

skibinalexander committed:
IOS-5588 Fix for pending transaction

skibinalexander commented:
@tureck1y можно смотреть

skibinalexander committed:
IOS-5588 Merge branch 'IOS-5591-algorand-integration' into IOS-5588-make-algorand-network-layer
